### PR TITLE
Reviewing 0-size ideals/modules

### PR DESCRIPTION
**Pull request by malex984**
NOTE: id_Delete will also be used for dense matrices

ADD: added assum'es, esp. concerning corner cases
ADD: doxygen documentation for id* functions and sip_sideal structure
CHG: minor improvements

### CODE REVIEW

File: libpolys/polys/simpleideals.cc [hunk 1]
```diff
@@ -59,10 +60,10 @@ ideal idInit(int idsize, int rank)
 void idShow(const ideal id, const ring lmRing, const ring tailRing, const int debugPrint)
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: No: idShow is a debug routine and must therefore accept id==NULL -> no need to change this routine

File: libpolys/polys/simpleideals.cc [hunk 2]
```diff
@@ -82,7 +83,13 @@ void idShow(const ideal id, const ring lmRing, const ring tailRing, const int de
 int id_PosConstant(ideal id, const ring r)
 {
   id_Test(id, r);
+  assume( id != NULL );
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: not needed: id_Test should do this.

File: libpolys/polys/simpleideals.cc [hunk 3]
```diff
@@ -82,7 +83,13 @@ void idShow(const ideal id, const ring lmRing, const ring tailRing, const int de
...
+
   const int N = IDELEMS(id) - 1;
+
+  if( id->m == NULL || N <= 0 )
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: not needed, the for loop is simply empty (and correct) in this case

File: libpolys/polys/simpleideals.cc [hunk 4]
```diff
@@ -96,60 +103,65 @@ int id_PosConstant(ideal id, const ring r)
   return -1;
 }
 
-/*2
-* initialise the maximal ideal (at 0)
-*/
+/// initialise the maximal ideal (at 0)
 ideal id_MaxIdeal (const ring r)
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: not needed (is only a reformulation)

File: libpolys/polys/simpleideals.cc [hunk 5]
```diff
@@ -96,60 +103,65 @@ int id_PosConstant(ideal id, const ring r)
...
   }
+  id_Test(hh, r);
   return hh;
 }
 
-/*2
-* deletes an ideal/matrix
-*/
+/// deletes an ideal/module/matrix
 void id_Delete (ideal * h, ring r)
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: dto: only a reformulation

File: libpolys/polys/simpleideals.cc [hunk 6]
```diff
@@ -161,14 +173,15 @@ void id_ShallowDelete (ideal *h, ring r)
   *h=NULL;
 }
 
-/*2
-*gives an ideal the minimal possible size
-*/
+/// gives an ideal/module the minimal possible size
 void idSkipZeroes (ideal ide)
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: dto: only a reformulation

File: libpolys/polys/simpleideals.cc [hunk 7]
```diff
@@ -210,25 +226,30 @@ int idElem(const ideal F)
...
 
-/*2
-* ideal id = (id[i])
-* result is leadcoeff(id[i]) = 1
-*/
+/// ideal id = (id[i]), result is leadcoeff(id[i]) = 1
 void id_Norm(ideal id, const ring r)
 {
+  id_Test(id, r);
+  assume (id != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should do this test

File: libpolys/polys/simpleideals.cc [hunk 8]
```diff
@@ -238,12 +259,13 @@ void id_Norm(ideal id, const ring r)
...
-/*2
-* ideal id = (id[i]), c any unit
-* if id[i] = c*id[j] then id[j] is deleted for j > i
-*/
+/// ideal id = (id[i]), c any unit
+/// if id[i] = c*id[j] then id[j] is deleted for j > i
 void id_DelMultiples(ideal id, const ring r)
 {
+  id_Test(id, r);
+  assume (id != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should do this test

File: libpolys/polys/simpleideals.cc [hunk 9]
```diff
@@ -277,12 +299,13 @@ void id_DelMultiples(ideal id, const ring r)
...
-/*2
-* ideal id = (id[i])
-* if id[i] = id[j] then id[j] is deleted for j > i
-*/
+/// ideal id = (id[i])
+/// if id[i] = id[j] then id[j] is deleted for j > i
 void id_DelEquals(ideal id, const ring r)
 {
+  id_Test(id, r);
+  assume (id != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should do this test

File: libpolys/polys/simpleideals.cc [hunk 10]
```diff
@@ -301,11 +324,12 @@ void id_DelEquals(ideal id, const ring r)
...
 }
 
-//
-// Delete id[j], if Lm(j) == Lm(i) and both LC(j), LC(i) are units and j > i
-//
+/// Delete id[j], if Lm(j) == Lm(i) and both LC(j), LC(i) are units and j > i
 void id_DelLmEquals(ideal id, const ring r)
 {
+  id_Test(id, r);
+  assume (id != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should do this test

File: libpolys/polys/simpleideals.cc [hunk 11]
```diff
@@ -328,12 +352,13 @@ void id_DelLmEquals(ideal id, const ring r)
...
-//
-// delete id[j], if LT(j) == coeff*mon*LT(i) and vice versa, i.e.,
-// delete id[i], if LT(i) == coeff*mon*LT(j)
-//
+/// delete id[j], if LT(j) == coeff*mon*LT(i) and vice versa, i.e.,
+/// delete id[i], if LT(i) == coeff*mon*LT(j)
 void id_DelDiv(ideal id, const ring r)
 {
+  id_Test(id, r);
+  assume (id != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should do this test

File: libpolys/polys/simpleideals.cc [hunk 12]
```diff
@@ -379,28 +404,29 @@ void id_DelDiv(ideal id, const ring r)
...
-*test if the ideal has only constant polynomials
-*/
+/// test if the ideal has only constant polynomials
+/// NOTE: zero ideal/module is also constant
 BOOLEAN id_IsConstant(ideal id, const ring r)
 {
-  int k;
-  for (k = IDELEMS(id)-1; k>=0; k--)
+  id_Test(id, r);
+  assume (id != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should do this test

File: libpolys/polys/simpleideals.cc [hunk 13]
```diff
@@ -379,28 +404,29 @@ void id_DelDiv(ideal id, const ring r)
...
+  
+  for (int k = IDELEMS(id)-1; k>=0; k--)
   {
     if (!p_IsConstantPoly(id->m[k],r))
       return FALSE;
   }
-  return TRUE;
+  return TRUE;
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: ??

File: libpolys/polys/simpleideals.cc [hunk 14]
```diff
@@ -379,28 +404,29 @@ void id_DelDiv(ideal id, const ring r)
...
-* copy an ideal
-*/
+/// copy an ideal
 ideal id_Copy(ideal h1, const ring r)
 {
-  int i;
-  ideal h2 =idInit(IDELEMS(h1),h1->rank);
-  for (i=IDELEMS(h1)-1; i>=0; i--)
+  id_Test(h1, r);
+  assume (h1 != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should do this test

File: libpolys/polys/simpleideals.cc [hunk 15]
```diff
@@ -408,25 +434,38 @@ ideal id_Copy(ideal h1, const ring r)
...
     omCheckAddrSize(h1,sizeof(*h1));
+
+    assume( h1->ncols >= 0 );
+    assume( h1->nrows >= 0 ); // matrix case?
+    
+    assume( h1->rank >= 0 );
+
+    assume( !( (h1->ncols*h1->nrows) > 0 && h1->m == NULL) );
+
     omdebugAddrSize(h1->m,h1->ncols*h1->nrows*sizeof(poly));
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: This needs: if(h1->ncols*h1->nrows >0) ....

File: libpolys/polys/simpleideals.cc [hunk 16]
```diff
@@ -471,6 +510,9 @@ static int p_Comp_RevLex(poly a, poly b,BOOLEAN nolex, const ring R)
 // uses lex-ordering when nolex = FALSE
 intvec *id_Sort(const ideal id, const BOOLEAN nolex, const ring r)
 {
+  id_Test(id, r);
+  assume (id != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_test should test this

File: libpolys/polys/simpleideals.cc [hunk 17]
```diff
@@ -561,43 +603,48 @@ intvec *id_Sort(const ideal id, const BOOLEAN nolex, const ring r)
...
 
-/*2
-* concat the lists h1 and h2 without zeros
-*/
+/// concat the lists h1 and h2 without zeros
 ideal id_SimpleAdd (ideal h1,ideal h2, const ring R)
 {
-  int i,j,r,l;
-  ideal result;
+  assume (h1 != NULL && h2 != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should test this

File: libpolys/polys/simpleideals.cc [hunk 18]
```diff
@@ -609,16 +656,20 @@ BOOLEAN idInsertPoly (ideal h1, poly h2)
...
+ * - if zeroOk is true, then h2 will also be inserted when it is zero
+ * - if duplicateOk is true, then h2 will also be inserted when it is
+ *   already present in h1
+ * return TRUE iff h2 was indeed inserted
+ */
 BOOLEAN id_InsertPolyWithTests (ideal h1, const int validEntries,
   const poly h2, const bool zeroOk, const bool duplicateOk, const ring r)
 {
+  id_Test(h1, r);
+  assume (h1 != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_test should test this

File: libpolys/polys/simpleideals.cc [hunk 19]
```diff
@@ -640,38 +691,41 @@ BOOLEAN id_InsertPolyWithTests (ideal h1, const int validEntries,
...
 }
 
-/*2
-* h1 + h2
-*/
+/// h1 + h2
 ideal id_Add (ideal h1,ideal h2, const ring r)
 {
+  id_Test(h1, r);
+  assume (h1 != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_test should test this

File: libpolys/polys/simpleideals.cc [hunk 20]
```diff
@@ -640,38 +691,41 @@ BOOLEAN id_InsertPolyWithTests (ideal h1, const int validEntries,
...
+  id_Test(h2, r);
+  assume (h2 != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_Test should test this

File: libpolys/polys/simpleideals.cc [hunk 21]
```diff
@@ -640,38 +691,41 @@ BOOLEAN id_InsertPolyWithTests (ideal h1, const int validEntries,
...
 }
 
-/*2
-* h1 * h2
-*/
+/// h1 * h2
 ideal  id_Mult (ideal h1,ideal  h2, const ring r)
 {
+  id_Test(h1, r);
+  assume (h1 != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_test should test this

File: libpolys/polys/simpleideals.cc [hunk 22]
```diff
@@ -640,38 +691,41 @@ BOOLEAN id_InsertPolyWithTests (ideal h1, const int validEntries,
...
+  id_Test(h2, r);
+  assume (h2 != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: id_test should test this

File: libpolys/polys/simpleideals.cc [hunk 23]
```diff
@@ -640,38 +691,41 @@ BOOLEAN id_InsertPolyWithTests (ideal h1, const int validEntries,
...
-  if (j == 0)
-    hh = idInit(1,1);
-  else
-    hh=idInit(j,1);
-  if (h1->rank<h2->rank)
-    hh->rank = h2->rank;
-  else
-    hh->rank = h1->rank;
+  
+  ideal  hh = idInit(j, si_max( h2->rank, h1->rank ));
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: must be: ideal  hh = idInit(si_max(1,j), si_max( h2->rank, h1->rank ));

File: libpolys/polys/simpleideals.cc [hunk 24]
```diff
@@ -687,63 +741,63 @@ ideal  id_Mult (ideal h1,ideal  h2, const ring r)
...
+  
+  id_Compactify(hh,r);
+  return hh;
 }
 
-/*2
-*returns true if h is the zero ideal
-*/
+/// returns true if h is the zero ideal
 BOOLEAN idIs0 (ideal h)
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: this is only a reformaution: no need to change this routine

File: libpolys/polys/simpleideals.cc [hunk 25]
```diff
@@ -687,63 +741,63 @@ ideal  id_Mult (ideal h1,ideal  h2, const ring r)
...
 }
 
-/*2
-* return the maximal component number found in any polynomial in s
-*/
+/// return the maximal component number found in any polynomial in s
 long id_RankFreeModule (ideal s, ring lmRing, ring tailRing)
 {
-  long  j=0;
+  assume (s != NULL);
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: should be done by id_testTail

File: libpolys/polys/simpleideals.cc [hunk 26]
```diff
@@ -687,63 +741,63 @@ ideal  id_Mult (ideal h1,ideal  h2, const ring r)
...
       }
-    }
   }
-  return j;
+  
+  return j; //  return -1;
 }
 
+/// test whether this input has vectors among entries
 BOOLEAN idIsModule(ideal id, ring r)
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: this reformulation is wrong: result must be FALSO for no/only 0-entries

File: libpolys/polys/simpleideals.cc [hunk 27]
```diff
@@ -1056,18 +1109,15 @@ void id_Compactify(ideal id, const ring r)
   idSkipZeroes(id);
 }
 
-/*2
-* returns the ideals of initial terms
-*/
+/// returns the ideals of initial terms
 ideal id_Head(ideal h,const ring r)
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: no need to change this routine (only reformulation)

File: libpolys/polys/simpleideals.cc [hunk 28]
```diff
@@ -1721,9 +1770,15 @@ ideal id_ChineseRemainder(ideal *xx, number *q, int rl, const ring r)
 
 void id_Shift(ideal M, int s, const ring r)
 {
+  assume( M != NULL );
```
Comment (hannes14) [inline libpolys/polys/simpleideals.cc]: should be done by id_test

File: libpolys/polys/simpleideals.h [hunk 29]
```diff
@@ -10,6 +10,11 @@
 #include <polys/monomials/ring.h>
 #include <polys/matpol.h>
 
+/// The following sip_sideal structure has many different uses
+/// thoughout Singular. Basic use-cases for it are:
+/// * ideal/module: nrows = 1, ncols >=0 and rank:0 for ideals, rank>0 for modules
```
Comment (hannes14) [inline libpolys/polys/simpleideals.h]: wrong: rank=1 for ideals, rank >=0 for modules!

**Comment by malex984** [comment]
@hannes14  thanks! i removed unnecessary assumes but kept doxygen documentation in place.

**Comment by hannes14** [comment]
already merged

**Pull request closed without merging**